Add helper methods to locate actors that can be reached via a path.

RoosterDragon opened the pull request:
Fixes AI trying to attack units they can't reach, and other automated behaviours will now only consider reachable locations, e.g. if rearming. Fixes #17753

Test scenario:

Start a map with an AI on an island and park some naval units near to their base. Deploy a barracks onto their island but far away from their base and build some units.

On bleed: When the AI creates a squad from the land units it will try and attack the nearest units - the naval units.
This PR: The squad will attack the nearest reachable units - the barracks. The closer naval units are ignored because the land units can't move to them.

----

Previously, the ClosestTo and PositionClosestTo existed to perform a simple distance based check to choose the closest location from a choice of locations to a single other location. For some functions this is sufficient, but for many functions we want to then move between the locations. If the location selected is in fact unreachable (e.g. on another island) then we would not want to consider it.

We now introduce ClosestToIgnoringPath for checks where we don't care about a path existing, e.g. weapons hitting nearby targets. When we do care about paths, we introduce ClosestToWithPathFrom and ClosestToWithPathTo which will check that a path exists. The PathFrom check will make sure one of the actors from the list can make it to the single target location. The PathTo check will make sure the single actor can make it to one of the target locations. This difference allows us to specify which actor will be doing the moving. This is important as a path might exists for one actor, but not another. Consider two islands with a hovercraft on one and a tank on the other. The hovercraft can path to the tank, but the tank cannot path to the hovercraft.

We also introduce WithPathFrom and WithPathTo. These will perform filtering by checking for valid paths, but won't select the closest location.

By employing the new methods that filter for paths, we fix various behaviour that would cause actors to get confused. Imagine an islands map, by checking for paths we ensure logic will locate reachable locations on the island, rather than considering a location on a nearby island that is physically closer but unreachable. This fixes AI squad automation, and other automated behaviours such as rearming.

dnqbob commented on OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs:
```diff
@@ -90,7 +92,7 @@ public void Tick(Squad owner)
 			}
 
-			var leader = owner.Units.ClosestTo(owner.TargetActor.CenterPosition);
+			var leader = owner.Units.ClosestToWithPathTo(owner.TargetActor);
```
I think this line may fix part of #17327, by using pathfinding distance to the target, instead of previous geometric distance.

dnqbob replied:
As for the other part, we need to limit leader's locomotor or use other methods to prevent entire squad stuck on terrain that only leader can pass.

RoosterDragon replied:
This is still using the geometric distance to determine closeness, so it won't aid with that sadly.

RoosterDragon replied:
Have filed #20228 to cover it though.

PunkPun commented on OpenRA.Mods.Common/WorldExtensions.cs:
```diff
@@ -17,6 +17,114 @@ namespace OpenRA.Mods.Common
+				return Enumerable.Empty<Actor>();
+
+			var pathFinder = sourceActor.World.WorldActor.Trait<PathFinder>();
+			var locomotor = sourceActor.Trait<Mobile>().Locomotor;
```
we don't need to get `Mobile` twice

RoosterDragon replied:
Updated

PunkPun commented on OpenRA.Mods.Common/WorldExtensions.cs:
```diff
@@ -17,6 +17,114 @@ namespace OpenRA.Mods.Common
+				return null;
+
+			var pathFinder = sourceActor.World.WorldActor.Trait<PathFinder>();
+			var locomotor = sourceActor.Trait<Mobile>().Locomotor;
```
same here

RoosterDragon replied:
Updated

PunkPun reviewed (commented):
Otherwise code LGTM

I've ran a large AI game and noticed some bugs, I'm not sure if they are regressions

Boat can't group
<img width="546" alt="Screenshot 2022-08-28 at 22 44 07" src="https://user-images.githubusercontent.com/37534529/187092454-1fa8cd7f-3d41-4557-b736-077ce3a72e78.png">

boats try to attack an out of range target
<img width="563" alt="Screenshot 2022-08-28 at 22 47 28" src="https://user-images.githubusercontent.com/37534529/187092479-a0789724-85af-4652-b827-fae0cc601275.png">

Artilleries unable to attack move
<img width="195" alt="Screenshot 2022-08-28 at 22 48 09" src="https://user-images.githubusercontent.com/37534529/187092496-7a312a44-d88c-43c0-95d3-8c39fa051693.png">

submarines try to move towards visible ground enemy units
<img width="550" alt="Screenshot 2022-08-28 at 22 55 22" src="https://user-images.githubusercontent.com/37534529/187092512-033bb539-f636-4b8b-a337-27dd59a58379.png">

PunkPun commented:
note: in my testing everything went perfectly, I just listed the few exceptions I noticed

RoosterDragon commented:
Hmm - the fact boats will move to attack land at all is actually suspect. That should be prevented by the new checks.

That also reveals a wider issues - we might want things like boats to get close to land to attack things on it, but if the checks were working they wouldn't ever try and do this unless they happened to be in autotarget range. So instead of seeing if we can reach the target, we might want to see if we can reach a location close to the target instead. That would be more complicated.

Will need to investigate this some more.

RoosterDragon commented:
#20239 fixes things so boats now ignore the land units.

However it seems like a regression to not allow boats to attack units near the shore - we need to teach the AI that it can attack units it can't reach - they don't need to get to the unit - just close enough to be in weapons range. Plus revalidating the other scenarios @PunkPun reported.

dnqbob commented:
I think it is ok to ignore the land units, for we also have an Autotarget can get target while the squad attack moving

PunkPun reviewed (commented):
Did another game (rebased on bleed)

The artillery but happened again (ignore the light tank he's attacking the airfield)
<img width="1203" alt="Screenshot 2022-09-04 at 13 39 30" src="https://user-images.githubusercontent.com/37534529/188310034-5ba5dde8-dc7a-4328-8d24-aa03edcd3265.png">

Naval grouping seems to be broken. These submarines attacked a naval yard which was built right next to them and did nothing afterwards. When a new sub was built it went kill the enemy naval yard, after that it also went idle. Not that there were anymore targets in water afterwards
<img width="189" alt="Screenshot 2022-09-04 at 13 37 12" src="https://user-images.githubusercontent.com/37534529/188310168-9cc6c285-ba90-403f-885d-7620f0f61bef.png">

In the end, the winning AI was unable to finish the game. There were 4 other AI's which had only the sub pen / naval yard left as their last structure
<img width="1680" alt="Screenshot 2022-09-04 at 13 57 56" src="https://user-images.githubusercontent.com/37534529/188310330-5a9c46b1-24dd-4300-b761-f61e0d070f18.png">

Probably unrelated to this PR, but AI nuked its own forces. (nuke was launched by the yellow AI)
<img width="349" alt="Screenshot 2022-09-04 at 13 44 41" src="https://user-images.githubusercontent.com/37534529/188310381-1ce485a1-0ed2-4cb0-9bb0-921767e90274.png">

RoosterDragon commented:
Pushed a commit that attempts to solve the problem where the AI squad can't directly path to the unit, e.g. boats trying to attack a land unit, but can get close enough to attack, e.g. the land units are on the beach.

Code is godawful and the overall approach still needs some thinking, but this is a step forward. Marking as a draft because more work is needed here.

PunkPun commented:
I would imagine the proper solution would be to generate a circle around the target the size of weapon range and see if all if any of those cells can be reached.

dnqbob commented:
I think it could be better that split commit "[Add helper methods to locate actors that can be reached via a path.](https://github.com/OpenRA/OpenRA/pull/20227/commits/3c5103a71fc2cce0c346b8e7fa82f5ab1ebb0f12)" into another PR and we merge that first. AI squad control is a really messing thing that will takes much longer time on fixing and testing.

RoosterDragon commented:
With only the first commit, AI navy won't attack any land units. This is because they can't path to them. That would probably be worse than the current behaviour of trying to attack land units they can't reach. So we probably do want a solution that finds the balance - navy can attack land units, but only tries to attack land units within range rather than anything.

dnqbob commented:
> With only the first commit, AI navy won't attack any land units. This is because they can't path to them. That would probably be worse than the current behaviour of trying to attack land units they can't reach. So we probably do want a solution that finds the balance - navy can attack land units, but only tries to attack land units within range rather than anything.

I think it is OK, because "current behaviour of trying to attack land units they can't reach" can give us a huge lag while also stuck the navy on the shore. I personally prefer a smoother gameplay on this issue. I remember RA2 also make AI navy can only attack navy, it is possible that it is due to the same reason.

PunkPun commented:
It is not fine, see my previous comment

> In the end, the winning AI was unable to finish the game. There were 4 other AI's which had only the sub pen / naval yard left as their last structure

dnqbob commented:
But I think it is better than when situation happends like https://github.com/MustaphaTR/Romanovs-Vengeance/issues/88, which make the game almost unplayable when there is too many unit trying to attack unreachable.

Add: for optimazing this situation I did many works, like only allow leader do the pathfinding to target, kick stuck unit from squad, but it also happen sometimes when units try to attack unreachable units.

anvilvapre commented:
Sounds like two different types of path searches, one to reach a destination, another to try to attack something (that will be) in target range?

dnqbob commented:
Could you please add `BlockedByActor` parameter to those pathfinder helper? To solve the problem of AI stuck by enclosed buildings we need a helper method like that.

RoosterDragon commented:
We're reliant on `HierarchicalPathFinder.PathExists` for this check. It isn't sufficiently smart enough to help with that I think. It can do `BlockedByActor.None` and some of `BlockedByActor.Immovable` but that's it.

I think something like #20945 is along the right lines anyway. Rather than trying to get clever and working out if placing a building will complete a wall that traps in some units, just enforcing spacing on some/all buildings help reduce the chance of congested AI bases that might trap things in.

RoosterDragon commented:
I have updated the code to deal with issues noted previously. Here are the scenarios which are now improved.

I used the following map for testing these scenarios out. Human player should use spawn 'A'. Enabling cheats and actor tags are useful. [IslandsPathfinding.zip](https://github.com/OpenRA/OpenRA/files/12087882/IslandsPathfinding.zip)

### Target Units on Same Island

- Check the AI squad target.
- Build a barracks in the bottom-right of your island.
- Build a barracks at the top of the AI island.

On bleed, the squad will initially target your base - the "Assault, fact XX" in the actor tags tells you the squad target. Once the first barracks is placed, they will target that instead - "tent XX". Once the second barracks is placed, they ignore it as the first is physically closer.

With this PR, they won't target the base as it isn't reachable. Same with the first barracks. Once the second barrack is placed, the squad will attack it.

![1](https://github.com/OpenRA/OpenRA/assets/3399086/e50ecb7c-f504-49f2-a0ea-8514a87d2c44)
With this PR, the AI squad ignores the closer barracks to move to the one on their island.

### Target Naval Units In Range of Island

- Build a naval yard, then a gunboat.
- Move the gunboat to the AI island shoreline.
- Wait for the AI to react.
- Move the gunboat along the coast, near the shore. Force the AI squad to keep moving to keep up with it.
- Create a second gunboat and place it near the enemy shoreline.
- Move the first gunboat out of range to the left side of the map.

On bleed, the AI squad will attack the gunboat. But when it moves out of range, they won't retarget. They won't respond to new units as a result. With the PR they will drop the target when it moves out of range - and can retarget to the new unit immediately.

![2](https://github.com/OpenRA/OpenRA/assets/3399086/f2ccbb80-1b3e-4fe3-a989-4142cb063a35)
Squad targetting the boat.

### Target Lands Units In Range Of Island

- Build some rifles and send them to the stair step at the bottom of your island.
- Wait for the AI squad to react.
- Move the units out of range.

The AI squad will attack these units. On bleed when the units move away the squad won't retarget and will keep trying to attack. With this PR they will go idle and return to base.

![3](https://github.com/OpenRA/OpenRA/assets/3399086/283afc89-de09-4dcd-bcf3-1961c4eb9da6)
The AI squad has dropped the targets that moved out of range.

### AI Units used after squad disbanded

- Set SquadManagerBotModuleInfo.RushInterval to 0 and SquadManagerBotModuleInfo.MinimumAttackForceDelay to 60000. This will ensurce the AI creates a Rush squad.
- Build a Construction Yard on the AI island.
- Wait for the AI squad to destroy it and return to base.
- Build another Construction Yard on the AI island.

On bleed, the squad stays put - they have retargetted to the unreachable construction yard and will be useless. With this PR the units will go idle. Once the next strucuture is placed a new squad is formed once the AI has enough units and will attack your new structure. Importantly, idle units from the first squad will be included in the force.

There existed a bug where an AI squad that disbands doesn't mark those units as available. So the AI never notices they can be assigned into another squad. So the units that returned to base from the first squad would never have been assigned to the new squad without this fix.

----

Finally, I tested a large AI game and didn't notice any performance issues. If you experience any, they are likely to present as lag spikes spaced approx 1 second apart. This is because the AI does some thinking approx every 1 second so it would almost certainly be AI related. If you experience this please include a replay! And if possible some repro steps.

PunkPun commented on OpenRA.Mods.Common/WorldExtensions.cs:
```diff
@@ -11,12 +11,191 @@
+		{
+			if (sourceActor.TraitOrDefault<Aircraft>() != null)
+				return actors.Select<Actor, (Actor Actor, WVec[] ReachableOffsets)>(a => (a, targetOffsets(a)));
+			var mobile = sourceActor.TraitOrDefault<Mobile>();
```
```suggestion
			if (sourceActor.Info.HasTraitInfo<AircraftInfo>())
				return actors.Select<Actor, (Actor Actor, WVec[] ReachableOffsets)>(a => (a, targetOffsets(a)));
			
			var mobile = sourceActor.TraitOrDefault<Mobile>();
```

If we don't use the variable it's better to retrieve the boolean directly

RoosterDragon replied:
Updated all.

PunkPun commented on OpenRA.Mods.Common/WorldExtensions.cs:
```diff
@@ -11,12 +11,191 @@
+		/// </summary>
+		public static WPos? ClosestToWithPathFrom(this IEnumerable<WPos> positions, Actor sourceActor)
+		{
+			if (sourceActor.TraitOrDefault<Aircraft>() != null)
```
same here

PunkPun commented on OpenRA.Mods.Common/WorldExtensions.cs:
```diff
@@ -11,12 +11,191 @@
+			return actors
+				.Where(a =>
+				{
+					if (a.TraitOrDefault<Aircraft>() != null)
```
and here

PunkPun commented on OpenRA.Mods.Common/WorldExtensions.cs:
```diff
@@ -11,12 +11,191 @@
+			return actors
+				.Select<Actor, (Actor Actor, WPos[] ReachablePositions)>(a =>
+				{
+					if (a.TraitOrDefault<Aircraft>() != null)
```
and here

PunkPun commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -925,6 +925,9 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 			{
+				if (!order.Target.IsValidFor(self))
+					return;
+
```
In a lot of cases these checks aren't necessary as target will be of terrain type making it always true, though I don't oppose the addition of sanity checks. Could they be pushed into their own commit though?

RoosterDragon replied:
Made a separate commit.

I had too add this check for AttackMove as the change in the bot code to use that instead of Attack was crashing, so decided to add all missing checks. As you say most of them aren't stricly required but seems better to have them everywhere than to not.

PunkPun commented on OpenRA.Mods.Common/Traits/CaptureManager.cs:
```diff
@@ -221,14 +221,19 @@ public bool StartCapture(Actor self, Actor target, CaptureManager targetManager,
 			if (enterMobile != null && enterMobile.IsMovingBetweenCells)
 				return false;
 
+			WPos? closestTargetablePosition = null;
+			if (move != null && captures.Info.ConsumedByCapture)
+			{
+				closestTargetablePosition = target.GetTargetablePositions().ClosestToWithPathFrom(self);
+				if (closestTargetablePosition == null)
+					return false;
+			}
+
```
why was this code moved outside? Afaik this code shouldn't care about pathfinding, the actor is moving just 1 cell distance into a cell that is likely blocked

RoosterDragon replied:
I was wondering what to do if the `closestTargetablePosition` came back as null.

If we can trust it, we don't need this. Reverted the changes and used `ClosestToIgnoringPath` instead.

PunkPun commented on OpenRA.Mods.Common/Traits/AutoCrusher.cs:
```diff
@@ -63,7 +63,7 @@ void INotifyIdle.TickIdle(Actor self)
 					Info.TargetRelationships.HasRelationship(self.Owner.RelationshipWith(a.Owner)) &&
 					a.TraitsImplementing<ICrushable>().Any(c => c.CrushableBy(a, self, Info.CrushClasses)))
-				.ClosestTo(self); // TODO: Make it use shortest pathfinding distance instead
+				.ClosestToWithPathFrom(self); // TODO: Make it use shortest pathfinding distance instead
```
Comment can be removed?

RoosterDragon replied:
It's still not the shortest path, so have left it for now. We're doing "closest to as the crow flies" but excluding units where a path doesn't exist. Consider the following contrived scenario:
![image](https://github.com/OpenRA/OpenRA/assets/3399086/c3d89bb1-df8f-4bb5-91b1-3dfbb5e25971)

`ClosestToWithPathFrom` will choose A as closest to start, it won't choose b which has a shorter path.

PunkPun commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -47,7 +47,7 @@ public static Actor ChooseResupplier(Actor self, bool unreservedOnly)
 					&& rearmInfo.RearmActors.Contains(a.Info.Name)
 					&& (!unreservedOnly || Reservable.IsAvailableFor(a, self)))
-				.ClosestTo(self);
+				.ClosestToWithPathFrom(self);
```
this activity is aircraft exclusive so it doesn't make sense to pathfind

RoosterDragon replied:
`ClosestToWithPathFrom` has knowledge that aircraft can path anywhere. So I think I'll keep this as-is so the special casing of Aircraft remains contained within the helpers.

RoosterDragon committed:
Validate order targets when resolving orders.

RoosterDragon committed:
Add helper methods to locate actors that can be reached via a path.

Previously, the ClosestTo and PositionClosestTo existed to perform a simple distance based check to choose the closest location from a choice of locations to a single other location. For some functions this is sufficient, but for many functions we want to then move between the locations. If the location selected is in fact unreachable (e.g. on another island) then we would not want to consider it.

We now introduce ClosestToIgnoringPath for checks where we don't care about a path existing, e.g. weapons hitting nearby targets. When we do care about paths, we introduce ClosestToWithPathFrom and ClosestToWithPathTo which will check that a path exists. The PathFrom check will make sure one of the actors from the list can make it to the single target location. The PathTo check will make sure the single actor can make it to one of the target locations. This difference allows us to specify which actor will be doing the moving. This is important as a path might exists for one actor, but not another. Consider two islands with a hovercraft on one and a tank on the other. The hovercraft can path to the tank, but the tank cannot path to the hovercraft.

We also introduce WithPathFrom and WithPathTo. These will perform filtering by checking for valid paths, but won't select the closest location.

By employing the new methods that filter for paths, we fix various behaviour that would cause actors to get confused. Imagine an islands map, by checking for paths we ensure logic will locate reachable locations on the island, rather than considering a location on a nearby island that is physically closer but unreachable. This fixes AI squad automation, and other automated behaviours such as rearming.

Mailaender commented:
Needs a rebase.

PunkPun commented on OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs:
```diff
@@ -194,22 +194,105 @@ void IBotTick.BotTick(IBot bot)
+				.Select(x => (x.Actor, Range: x.AttackBases.Max(ab => ab.GetMaximumRangeVersusTarget(target))))
+				.ToDictionary(x => x.Actor, x => x.Range);
+
+			// Now determine if each actor can either path directly to the target,
+			// or if it can path to a nearby location at the edge of its weapon range to the target
+			// A thorough check would check each position within the circle, but for performance
+			// we'll only check 8 positions around the edge of the circle.
+			// We need to account for the weapons range here to account for units such as boats.
+			// They can't path directly to a land target,
+			// but might be able to get close enough to shore to attack the target from range.
+			return ownActorsAndTheirAttackRanges.Keys
+				.ClosestToWithPathToAny(targetActor.World, a =>
+				{
+					var range = ownActorsAndTheirAttackRanges[a].Length;
+					var rangeDiag = Exts.MultiplyBySqrtTwoOverTwo(range);
+					return new[]
+					{
+						targetActor.CenterPosition,
+						targetActor.CenterPosition + new WVec(range, 0, 0),
+						targetActor.CenterPosition + new WVec(-range, 0, 0),
+						targetActor.CenterPosition + new WVec(0, range, 0),
+						targetActor.CenterPosition + new WVec(0, -range, 0),
+						targetActor.CenterPosition + new WVec(rangeDiag, rangeDiag, 0),
+						targetActor.CenterPosition + new WVec(-rangeDiag, rangeDiag, 0),
+						targetActor.CenterPosition + new WVec(-rangeDiag, -rangeDiag, 0),
+						targetActor.CenterPosition + new WVec(rangeDiag, -rangeDiag, 0),
```
I'm still unsure how to feel about this change. I feel some sort of backup system would still be needed to account for edge cases

RoosterDragon replied:
Can you expand on this? Specifically what edge cases do you think are possible and what sort of backup would we need to employ if they occured?

PunkPun replied:
Edge cases could be something like a river, more chocky areas, or just obstacles miraculously blocking all the 9 cells without actually blocking the target

RoosterDragon replied:
If an unfortunate terrain layout means a unit can't be targetted, then yes this will (incorrectly) ignore that unit. If it's the only possible option then indeed we would ignore it in this edge case.

However the following mitigate this scenario in my mind:
- If there's other units, the squad can attack those instead.
- If there's another squad with a different attack range, it might not suffer from the same unfortunate terrain setup and be able to attack.
- If there's another squad in the same domain (e.g. they are on the same island as the target) they can attack the target once available.
- If there's aircraft, they can mop up this target.

The required adverse terrain setup is quite specific that it should be very rare in practise, and if it does occur I think any of the above options mean it shouldn't matter really.

PunkPun commented on OpenRA.Mods.Common/Activities/Hunt.cs:
```diff
@@ -36,7 +36,7 @@ public override bool Tick(Actor self)
 				return true;
 
-			var targetActor = targets.ClosestTo(self);
+			var targetActor = targets.ClosestToWithPathFrom(self);
```
hunt will also need distance checks so that  targets on different terrain don't become invalid

RoosterDragon replied:
Since it's not caching the target I believe it is always choosing a fresh, valid target. I'm not sure what you'd want me to change here.

PunkPun commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/059c47dfa1e3a0b3d71461c7e587b2a59461580b...25abdd13a24478592a65685fc6f4dbbf2e496a23)
